fix(angular): Combobox clrMulti input handle string value of 'false'

IrisTien opened the pull request:
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)
- [ ] If applicable, have a visual design approval

## PR Type

What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] clarity.design website / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: #6432

## What is the new behavior?
Combobox Component clrMulti input handles both string and boolean type values.
If clrMulti input is boolean, it will pass the boolean value directly;
If clrMulti input is string: it would treat "false" and empty string to boolean false, it would set to true for all the other strings.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->

## Other information

gnomeontherun commented on projects/angular/src/forms/combobox/combobox.ts:
```diff
@@ -194,7 +194,14 @@ export class ClrCombobox<T>
   @Input('clrMulti')
   set multiSelect(value: boolean | string) {
-    if (value) {
+    let isMultiSelectModal = false;
```
I'm not sure in what situation you would be passing a string here. This is meant to be a boolean, and generally its good to restrict typing in these cases. It can lead to confusing outcomes for consumers if they think a string can be passed like `"0"`, is sending a boolean really a problem?

Regardless, this is a lot of logic. We can just coerce the value into a boolean and not try to handle individual type cases.

```js
value = !!value;
```

Jinnie replied:
On our website the usage is documented as:
> The multi selection mode is triggered with a clrMulti="true" 

This implies that `clrMulti="false"` should also work.
And I don't thing that the coercion will not work with the case of `clrMulti="false"`

mathisscott replied:
This would be better checking against more explicit values. There's a lot of unnecessary `typeof` checks here that are only necessary because we are duck typing the value.

So something more like...

```
let isMultiSelectModal: boolean;

switch (true) {
  case (value === 'false' || value === false):
     isMultiSelectModal = false;
     break;
  case (value === 'true' || value === true):
     isMultiSelectModal = true;
     break;
   default:
     isMultiSelectModal = false;
}
```

... then you could finish off with something like...

```
this.optionSelectionService.selectionModel = isMultiSelectModal ? new MultiSelectComboboxModel<T>() : new SingleSelectComboboxModel<T>();
```

...narrowing down a lot of conditional code here.

IrisTien replied:
Please check the comments from Bozhidar Dryanovski in issue #6432 :
`The reason for having string as a type is that the strict Angular (fullTemplateTypeCheck & strictTemplates) check requires every field that is boolean to have an additional type of string.`

So here we have the fix to support string value 'true' and 'false'.

@mathisscott , addressed the comments.

Jinnie commented:
Hello, @IrisTien,

There is a small formatting error that needs to be fixed. And also we need those 4 commits squashed, and then we can merge this fix in.

You can easily accomplish both tasks by soft resetting to the latest "good" commit, then running format:fix on the staged commits and force pushing all back. These are the commands you need:
```
git reset --soft 5911f81caae3e02ac9b9008c0d82bf07ad040529
npm run format:fix
```
Then commit (don't forget the "signed-off-by" line) and force push back to your fork. This PR will be automatically updated and build tasks rerun.

Thanks!

Commit:
fix(angular): Combobox clrMulti input handle string value of 'false'

- Add type checking for input clrMulti for Combobox Component. If the input type is string, it converts 'false' to boolean false.
- Add relative UT

Close: #6432

Signed-off-by: Wei Tian <twei@vmware.com>

ashleyryan commented:
Hello, this repository has been archived. Please visit one of the new repositories:

- [Clarity Core](https://github.com/vmware-clarity/core)
- [Clarity Angular](https://github.com/vmware-clarity/ng-clarity)
